filebeat/harvester/log_file.go

```diff
@@ -57,40 +61,49 @@ func (r *LogFile) Read(buf []byte) (int, error) {
 			r.offset += int64(n)
 			r.lastTimeRead = time.Now()
 		}
+		totalN += n
 
-		// reset backoff
+		// Read from source completed without error
+		// Either end reached or buffer full
 		if err == nil {
+			// reset backoff for next read
 			r.backoff = r.config.Backoff
-			return n, nil
+			return totalN, nil
 		}
 
-		if err != io.EOF {
-			logp.Err("Unexpected state reading from %s; error: %s", r.fs.Name(), err)
-			return n, err
-		}
+		// Move buffer forward for next read
+		buf = buf[n:]
 
-		// Stdin is not continuable
-		if !r.fs.Continuable() {
-			logp.Debug("harvester", "Source is not continuable: %s", r.fs.Name())
-			return n, err
+		// Buffer is full. Cannot continue reading. Return.
+		// Can happen if n == bufferSize + io.EOF error
+		if len(buf) == 0 {
+			return totalN, err
```

Comment:
we still need err to by rewritten by r.errorChecks before returning?

btw. in io.Reader interface it's valid to postpone error and return the `0, err` with next call to read.